driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -187,445 +144,603 @@ private void logError(InetSocketAddress address, Throwable exception) {
-                            write(connection, prepareAndRetry(toPrepare.getQueryString()));
-                            // we're done for now, the prepareAndRetry callback will handle the rest
-                            return;
-                        default:
-                            if (metricsEnabled())
-                                metrics().getErrorMetrics().getOthers().inc();
-                            break;
-                    }
+        void sendRequest() {
+            try {
+                while (!isDone.get() && queryPlan.hasNext() && !queryStateRef.get().isCancelled()) {
+                    Host host = queryPlan.next();
+                    logger.trace("[{}] Querying node {}", id, host);
+                    if (query(host))
+                        return;
+                }
+                reportNoMoreHosts(this);
+            } catch (Exception e) {
+                // Shouldn't happen really, but if ever the loadbalancing policy returned iterator throws, we don't want to block.
+                setFinalException(null, new DriverInternalError("An unexpected error happened while sending requests", e));
+            }
+        }
 
-                    if (retry == null)
-                        setFinalResult(connection, response);
-                    else {
-                        switch (retry.getType()) {
-                            case RETRY:
-                                ++retriesByPolicy;
-                                if (logger.isDebugEnabled())
-                                    logger.debug("Doing retry {} for query {} at consistency {}", retriesByPolicy, statement, retry.getRetryConsistencyLevel());
-                                if (metricsEnabled())
-                                    metrics().getErrorMetrics().getRetries().inc();
-                                retry(true, retry.getRetryConsistencyLevel());
-                                break;
-                            case RETHROW:
-                                setFinalResult(connection, response);
-                                break;
-                            case IGNORE:
-                                if (metricsEnabled())
-                                    metrics().getErrorMetrics().getIgnores().inc();
-                                setFinalResult(connection, new Responses.Result.Void());
-                                break;
-                        }
-                    }
-                    break;
-                default:
-                    setFinalResult(connection, response);
-                    break;
+        private boolean query(final Host host) {
+            currentPool = manager.pools.get(host);
+            if (currentPool == null || currentPool.isClosed())
+                return false;
+
+            if (nextExecutionScheduled.compareAndSet(false, true))
+                scheduleExecution(speculativeExecutionPlan.nextExecution(host));
```

Comment:
The problem with speculative retries is that they happen before we've received a response from the server, so we might not know anything about the statement (for example if it's a SimpleStatement, since we don't parse query strings client-side).
That's the reason why I introduced `StatementWrapper`. Users who want more precise control would need to "mark" the relevant statements and have special cases in a custom policy.
I don't see a way to provide that in a generic way in the driver.